### PR TITLE
Log transfer server and process server

### PR DESCRIPTION
**Pull request by MateusPedrosa**
Depends on:
- [ ] https://github.com/rock-core/package_set/pull/191

We created a module for spawning a log transfer FTP server and added the remote_processes directory from https://github.com/rock-core/tools-orocosrb/tree/master/lib/orocos/remote_processes, adding to it a new command for uploading log files to the FTP server and removed the deprecated COMMAND_MOVE_LOG.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -21,4 +21,8 @@ coverage/
 .yardoc
 
 
-tmp/
\ No newline at end of file
+tmp/
+
+lib/syskit/roby_app/log_transfer_server/Gemfile
+lib/syskit/roby_app/log_transfer_server/build_syskit.sh
+syskit-0.1.gem
```
Comment (doudou) [inline .gitignore:28]: Fairly sure this is not necessary anymore

File: lib/syskit/roby_app/log_transfer_server/spawn_server.rb [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                    @password = password
+                    @account = account
+                    @session_timeout = session_timeout
+                    @nat_ip = nat_ip
+                    @passive_ports = passive_ports
+                    @debug = debug
+                    @log = log
+                    @driver = Driver.new(@user, @password, @account, @data_dir)
+                    @server = Ftpd::FtpServer.new(@driver)
+                    configure_server
```
Comment (doudou) [inline lib/syskit/roby_app/log_transfer_server/spawn_server.rb:48]: I'm not sure there's a lot of value in storing every attribute as instance variable inside the class. You basically store them here, and the only thing you use them for is to setup the FTP server. Just store what really has to be stored, which from what I see is `@debug`

~~~
driver = Driver.new(user, password, account, data_dir)
server = Ftpd::FtpServer.new(driver)
server.interface = interface
server.port = port
server.tls = tls
server.passive_ports = passive_ports
server.certfile_path = certfile_path
server.auth_level = Ftpd.const_get("AUTH_#{@auth_level.upcase}")
server.session_timeout = session_timeout
server.log = make_log
server.nat_ip = nat_ip
@server = server
@server.start
~~~

File: lib/syskit/roby_app/log_transfer_server/spawn_server.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                end
+
+                def connection_host
+                    addr = IPAddr.new(@server.interface)
+                    if addr.ipv6?
+                        "::1"
+                    else
+                        "127.0.0.1"
+                    end
+                end
```
Comment (doudou) [inline lib/syskit/roby_app/log_transfer_server/spawn_server.rb:114]: What is that for ?

**Comment by planthaber** [comment]
I think this is great for automatic uploading to the ftp.

As a quick tip: When I only need a quick look from another host I often use `$> python -m SimpleHTTPServer` in  the logfile folder.

After that you can use a browser to see the folder contents or use `wget --recursive` to download the logs.

**Comment by doudou** [comment]
@MateusPedrosa I also had this comment in the previous pull request: "Also, you need to change https://github.com/rock-core/tools-syskit/blob/master/lib/syskit/scripts/process_server.rb to use the new code."

**Comment by MateusPedrosa** [comment]
> @MateusPedrosa I also had this comment in the previous pull request: "Also, you need to change https://github.com/rock-core/tools-syskit/blob/master/lib/syskit/scripts/process_server.rb to use the new code."

That's right, I forgot. Should I also do something about syskit/roby_app/process_server.rb or just leave it there?

**Comment by doudou** [comment]
> Should I also do something about syskit/roby_app/process_server.rb or just leave it there?

Should remove it.

**Comment by doudou** [comment]
To make sure we don't still have references to the orocos.rb process server, the simplest would be to
- add a "raise" at the top of lib/orocos/remote_processes/server.rb and lib/orocos/remote_processes/client.rb
- go into bundles/wetpaint and run `syskit run -rgazebo -c`

File: lib/syskit/roby_app/log_transfer_server/driver.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
+# !/usr/bin/env ruby
```
Comment (doudou) [inline lib/syskit/roby_app/log_transfer_server/driver.rb:1]: This is only useful if you execute the script

File: lib/syskit/roby_app.rb [hunk 5]
```diff
@@ -20,6 +20,8 @@ module RobyApp
 require "syskit/roby_app/single_file_dsl"
 require "syskit/roby_app/unmanaged_process"
 require "syskit/roby_app/unmanaged_tasks_manager"
+require 'syskit/roby_app/remote_processes'
+require 'syskit/roby_app/log_transfer_server'
```
Comment (doudou) [inline lib/syskit/roby_app.rb:24]: This will fail on rubocop

**Comment by doudou** [comment]
The ftpd dependency is missing. You need to declare it in the osdeps files of [rock-core](https://github.com/rock-core/package_set/) - checked out in autoproj/remotes/rock.core - and add the dependency in syskit's `manifest.xml`.

Name the new package `ruby-ftpd` to avoid having a much too generic `ftpd`. [Documentation](https://www.rock-robotics.org/rock-and-syskit/workspace/os_dependencies.html)

You will need to declare the dependency between the two pull requests. To do so, change the toplevel comment for *this* pull request to add

~~~
Depends on:
- [ ] URL_OF_THE_ROCK_CORE_PULL_REQUEST
~~~

File: .gitignore [hunk 6]
```diff
@@ -22,3 +22,7 @@ coverage/
 
 
 tmp/
+syskit-0.1.gem
+test/roby_app/spawn_server/Gemfile
+test/roby_app/spawn_server/build_syskit(1).sh
+test/roby_app/spawn_server/testfile
```
Comment (doudou) [inline .gitignore:28]: ?

File: lib/syskit/roby_app/log_transfer_server/write_only_disk_file_system.rb [hunk 7]
```diff
@@ -1,7 +1,7 @@
 # frozen_string_literal: true
 
-require "ftpd"
-require "pathname"
+# require "ftpd"
+# require "pathname"
```
Comment (doudou) [inline lib/syskit/roby_app/log_transfer_server/write_only_disk_file_system.rb]: No leaving commented-out code

File: test/roby_app/spawn_server/test_spawn_server.rb [hunk 8]
```diff
@@ -35,13 +35,14 @@ def upload_log(host, port, certificate, user, password, localfile)
             ftp.login(user, password)
             lf = File.open(localfile)
             ftp.storbinary("STOR #{File.basename(localfile)}", lf, Net::FTP::DEFAULT_BLOCKSIZE)
+            lf.close
```
Comment (doudou) [inline test/roby_app/spawn_server/test_spawn_server.rb]: In this particular case,

~~~
File.open(localfile) do |lf|
  ftp.storbinary("STOR #{File.basename(localfile)}", lf, Net::FTP::DEFAULT_BLOCKSIZE)
end
~~~

is much better, as the file will be closed even if `storbinary` throws an exception

File: test/roby_app/spawn_server/test_spawn_server.rb [hunk 9]
```diff
@@ -35,13 +35,14 @@ def upload_log(host, port, certificate, user, password, localfile)
...
         end
     end
 
     def upload_testfile
-        File.new("testfile", "w+")
-        upload_log(@server.interface, @server.port, @server.certfile_path, @server.user, @server.password, "testfile")
-        File.delete("testfile")
+        tf = File.open('testfile', 'w')
+        upload_log("127.0.0.1", @server.port, @server.certfile_path, @server.user, @server.password, tf)
+        tf.close
```
Comment (doudou) [inline test/roby_app/spawn_server/test_spawn_server.rb:45]: Same comment. Use block form.

File: test/roby_app/spawn_server/test_spawn_server.rb [hunk 10]
```diff
@@ -62,29 +63,28 @@ def upload_testfile
...
+        #         ca_file: @server.certfile_path) do |ftp|
                 
-                ftp.login(@server.user, @server.password)
-                assert_raises(Net::FTPPermError) { ftp.get("#{@temp_dir}/testfile") }
-            end
-        end
+        #         ftp.login(@server.user, @server.password)
+        #         assert_raises(Net::FTPPermError) { ftp.get("#{@temp_dir}/testfile") }
+        #     end
+        # end
```
Comment (doudou) [inline test/roby_app/spawn_server/test_spawn_server.rb]: ?

**Comment by g-arjones** [comment]
@doudou Does it make sense that most of our orogen tests fail with this patch?

**Comment by doudou** [comment]
> @doudou Does it make sense that most of our orogen tests fail with this patch?

No. Don't worry, we noticed ;-)

**Comment by doudou** [comment]
@rbtmrcs: just checking that we're on the same page, we're still missing the "invalid certificate" test, right ?

**Comment by doudou** [comment]
Also, @rbtmrcs, please run "rubocop" before you push something. I finally managed to get CI to pass on this one, and your new commit fails on rubocop.

**Review by doudou — APPROVED** [review]

**Comment by planthaber** [comment]
What about: https://github.com/rock-core/package_set/pull/191

**Pull request merged**